python/samples/demos/guided_conversations/guided_conversation/guided_conversation_agent.py

```diff
@@ -0,0 +1,352 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+from collections.abc import Callable
+from dataclasses import dataclass, field
+from enum import Enum
+import logging
+
+from pydantic import BaseModel
+from semantic_kernel import Kernel
+
+from guided_conversation.functions.conversation_plan import conversation_plan_function
+from guided_conversation.functions.execution import end_conversation, execution, send_message
+from guided_conversation.functions.final_update_plan import final_update_plan_function
+from guided_conversation.plugins.agenda import Agenda
+from guided_conversation.plugins.artifact import Artifact
+from guided_conversation.utils.conversation_helpers import Conversation, ConversationMessage, ConversationMessageType
+from guided_conversation.utils.openai_tool_calling import (
+    ToolValidationResult,
+    parse_function_result,
+    validate_tool_calling,
+)
+from guided_conversation.utils.plugin_helpers import PluginOutput, format_kernel_functions_as_tools
+from guided_conversation.utils.resources import GCResource, ResourceConstraint
+
+MAX_DECISION_RETRIES = 2
+
+
+class ToolName(Enum):
+    UPDATE_ARTIFACT_TOOL = "update_artifact_field"
+    UPDATE_AGENDA_TOOL = "update_agenda"
+    SEND_MSG_TOOL = "send_message_to_user"
+    END_CONV_TOOL = "end_conversation"
+
+
+@dataclass
+class GCInput:
+    artifact: BaseModel
+    rules: list[str]
+    conversation_flow: str | None
+    context: str | None
+    resource_constraint: ResourceConstraint | None
+
+
+@dataclass
+class GCOutput:
+    """The output of the GuidedConversation agent.
+
+    Args:
+        ai_message (str): The message to send to the user.
+        is_conversation_over (bool): Whether the conversation is over.
+    """
+
+    ai_message: str | None = field(default=None)
+    is_conversation_over: bool = field(default=False)
+
+
+class GuidedConversation:
+    def __init__(self, gc_input: GCInput, kernel: Kernel, service_id: str = "gc_main") -> None:
+        """Initializes the GuidedConversation agent.
+
+        Args:
+            gc_input (GCInput): The input to the guided conversation.
+            kernel (Kernel): An instance of Kernel. Must come initialized with a AzureOpenAI or OpenAI service.
+            service_id (str): Provide a service_id associated with the kernel's service that was provided.
+        """
+
+        self.logger = logging.getLogger(__name__)
+        self.kernel = kernel
+        self.service_id = service_id
+        self.gc_input = gc_input
+
+        self.chat_history = Conversation()
```

Comment:
would prefer if this wasn't called chat_history as that might confuse people who might think it is a `ChatHistory` class object instead of a `Conversation` object.
```suggestion
        self.conversation = Conversation()
```

Reply:
Resolved